src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/FeatureSwitchManager.cs

```diff
@@ -139,9 +146,17 @@ public MethodIL GetMethodILWithInlinedSubstitutions(MethodIL method)
             // Last step is a sweep - we replace the tail of all unreachable blocks with "br $-2"
             // and nop out the rest. If the basic block is smaller than 2 bytes, we don't touch it.
             // We also eliminate any EH records that correspond to the stubbed out basic block.
+            //
+            // We also attempt to rewrite calls to SR.SomeResourceString accessors with string
+            // literals looked up from the managed resources.
 
             Debug.Assert(method.GetMethodILDefinition() == method);
 
+            // Do not attempt to inline resource strings if we only want to use resource keys.
+            // The optimizations are not compatible.
+            bool shouldInlineResourceStrings =
+                !_hashtable._switchValues.TryGetValue("System.Resources.UseSystemResourceKeys", out bool useResourceKeys) || !useResourceKeys;
```

Comment:
Should we enable this only under some feature switch - or have other ways to opt-in/out.
Maybe something internal if things go wrong?

I don't have a good sense of the risk associated with this. It's just that typically we have ways to disable relatively agreesive optimizations.

Reply:
We haven't been doing this for most other optimizations, unless they look risky (preinit comes to mind). Piping them through is annoying, unless we misuse the feature switch mechanism. This one doesn't feel particularly risky. We aren't doing any major structure changes.

When I look at RyuJIT - most of the small optimization like this don't have explicit opt outs either and many of those that do only have an opt out in debug/checked builds.